myna/core/files/file_grains.py

```diff
@@ -0,0 +1,229 @@
+    dfRefIds = pd.read_csv(
+        reference_id_filename, skiprows=1, header=None, names=col_names
+    )
+    num_reference_ids = len(dfRefIds)
+    misorientation_z_ref = np.zeros(num_reference_ids)
+    # Z components of the three unit vectors defining the crystal orientation
+    nz1 = abs(dfRefIds["nz1"])
+    nz2 = abs(dfRefIds["nz2"])
+    nz3 = abs(dfRefIds["nz3"])
+    # Calculate <100> misorientation with Z and append to dataframe
+    for i in range(num_reference_ids):
+        if (nz1[i] >= nz2[i]) and (nz1[i] >= nz3[i]):
+            misorientation_z_ref[i] = math.degrees(math.acos(nz1[i]))
+        elif (nz2[i] >= nz1[i]) and (nz2[i] >= nz3[i]):
+            misorientation_z_ref[i] = math.degrees(math.acos(nz2[i]))
+        else:
+            misorientation_z_ref[i] = math.degrees(math.acos(nz3[i]))
+    return misorientation_z_ref
+
+
+def get_misorientation_z(df, misorientation_z_ref):
+    """args:
+        df: dataframe containing gid values
+        misorientation_z_ref: np array of misorientation values, in degrees,
+        for each possible grain orientation
+
+    Returns:
+        misorientation_z_list: np array of misorientation values, in degrees,
+        for each cell in the df"""
+    # Absolute value of grain ID used for conversion
+    gid_abs = abs(df["gid"].values)
+    num_gid_values = len(gid_abs)
+    misorientation_z_list = np.zeros(num_gid_values)
+    num_reference_ids = len(misorientation_z_ref)
+    # gid_index can be replaced with the stored gid value
+    # from the df after indexing bug is fixed
+    for i in range(num_gid_values):
+        gid_index = int((gid_abs[i] - 1) % num_reference_ids)
+        misorientation_z_list[i] = misorientation_z_ref[gid_index]
+    return misorientation_z_list
+
+
+def get_bin_centers_fre_dia(df2D, misorientation_z_list):
+    """Freedman–Diaconis rule for bin width calculation based on the skewed
+    nature of both the untextured and actual misorientation distributions
+    Bin data based on the number of unique grains in the data
+    Args:
+        df: dataframe containing gid values
+        misorientation_z_list: list of grain misorientations with <100>, in
+        degrees, used to map gid values
+
+    Returns:
+        bin_edges: edges for binned misorientation data
+        bin_centers: center location of bins for misorientation data"""
+    num_grains = len(df2D["gid"].unique())
+    bin_width_ideal = 2 * iqr(misorientation_z_list) / (num_grains ** (1 / 3))
+    # Get edges of bins for misorientation data (0 to 54.7 degrees)
+    num_bins = round(54.7 / bin_width_ideal)
+    bin_width = 54.7 / num_bins
+    bin_edges = [None] * (num_bins + 1)
+    for i in range(num_bins + 1):
+        bin_edges[i] = i * bin_width
+    # Get bin centers
+    bin_centers = [None] * num_bins
+    for i in range(num_bins):
+        bin_centers[i] = bin_edges[i] + bin_width / 2
+    return [bin_edges, bin_centers]
```

Comment:
These are defined in `myna.application.exaca.grainstats.py` and should be removed from here.